File: Topics/Arrays/496.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        map<int, int> mapTwoVector;
        vector<int> stack;
        int n = nums2.size();

        for (int i = 0; i < n; i++) {
            while(!stack.empty() && nums2[i] > stack[stack.size() - 1]) {
                mapTwoVector[stack.back()] = nums2[i];
                stack.pop_back();
            }
            stack.push_back(nums2[i]);
        }
        int idx = 0;
        for (auto num: nums1) {
            int tmp = mapTwoVector[num];
            tmp ? nums1[idx] = tmp: nums1[idx] = -1 ;
            idx++;
        }
        return nums1;
    }
};

File: Topics/Stack/496.py
from typing import Optional, List

class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:

        stack = ans = []
        dic = {}

        for i in nums2:
            while stack and stack[-1] < i:
                dic[stack.pop()] = i
            stack.append(i)

        while stack:
            dic[stack.pop()] = -1

        # print(dic)
        for i in nums1:
            ans.append(dic[i])
        return ans

slv = Solution()
x = slv.nextGreaterElement([4,1,2], [1,2,4])
print(x)